hutch_python/namespace.py

```diff
@@ -41,6 +47,23 @@ def class_namespace(cls, scope=None):
                 logger.debug(exc, exc_info=True)
                 return class_space
 
+    cache = set()
+
+    # Helper function to recursively add subdevices to the scope
+    def accumulate(obj, scope_objs, cache):
+        if obj not in cache:
+            cache.add(obj)
+            for comp_name in getattr(obj, 'component_names', []):
+                sub_obj = getattr(obj, comp_name)
+                accumulate(sub_obj, scope_objs, cache)
+            # Don't accidentally override
+            if obj.name not in scope_objs:
+                scope_objs[obj.name] = obj
+
+    for name, obj in scope_objs.copy().items():
```

Comment:
Why copy?

Reply:
To avoid mutating a dictionary that we're iterating through